applications/finetune-quora-embeddings/embed.py

```diff
@@ -0,0 +1,284 @@
+from modal import Volume, Image, Stub, gpu, Secret
+import os
+from helpers.models import EmbeddingModel, model_types
+from datasets import Dataset
+
+# DATASET_CONFIG
+DATASET_NAME = "567-labs/cleaned-quora-dataset-train-test-split"
+DATASET_DIR = "/data"
+DATASET_VOLUME = Volume.persisted("datasets")
+CACHE_DIRECTORY = f"{DATASET_DIR}/cached-embeddings"
+
+# Model Configs
+MODELS = [
+    "BAAI/bge-base-en-v1.5",
+    "llmrails/ember-v1",
+    "thenlper/gte-large",
+    "infgrad/stella-base-en-v2",
+    "sentence-transformers/gtr-t5-large",
+]
+OPENAI_MODELS = ["text-embedding-3-small", "text-embedding-ada-002"]
+
+COHERE_MODELS = ["embed-multilingual-v3.0"]
+
+SEMAPHORE_LIMITS: dict[model_types, int] = {
+    "HuggingFace": 20,
+    "OpenAI": 20,
+    "Cohere": 20,
+}
+
+BATCH_SIZE_CONFIG: dict[model_types, int] = {
+    "HuggingFace": 10000,
+    "OpenAI": 64,
+    "Cohere": float("inf"),
+}
+
+GPU_CONFIG = gpu.A100()
+
+
+def download_model():
+    from sentence_transformers import SentenceTransformer
+
+    for model_name in MODELS:
+        print(f"Downloading and caching model: {model_name}")
+        SentenceTransformer(model_name)
+
+
+def determine_if_model_has_generated_embeddings(model_name):
+    model_id = model_name.split("/").pop() if model_name in MODELS else model_name
+    train_file_path = f"{CACHE_DIRECTORY}/{model_id}-train.arrow"
+    test_file_path = f"{CACHE_DIRECTORY}/{model_id}-test.arrow"
+    return os.path.exists(train_file_path) and os.path.exists(test_file_path)
+
+
+image = (
+    Image.debian_slim()
+    .env({"Random": "123"})
+    .pip_install(
+        "cohere",
+        "datasets",
+        "sentence-transformers",
+        "scikit-learn",
+        "tabulate",
+        "openai",
+        "pyarrow",
+        "tenacity",
+        "diskcache",
+    )
+    .run_function(download_model)
+)
+
+stub = Stub("embeddings")
+
+
+def get_unique_sentences(
+    test_dataset: Dataset, sentence_to_id_mapping: dict, batch_size=1000
+):
+    seen = set()
+    batch = []
+    for row in test_dataset:
+        s1, s2 = row["questions"]["text"]
+
+        if s1 not in seen:
+            sentence_to_id_mapping[s1] = len(seen)
+            seen.add(s1)
+            batch.append(s1)
+
+            if len(batch) == batch_size:
+                yield batch
+                batch = []
+
+        if s2 not in seen:
+            sentence_to_id_mapping[s2] = len(seen)
+            seen.add(s2)
+            batch.append(s2)
+
+            if len(batch) == batch_size:
+                yield batch
+                batch = []
+
+    if batch:
+        yield batch
+
+
+def update_dataset_with_embeddings(
```

Comment:
i feel like a lot of functions could be organized better into 

`{directory}/{main.py|utils.py}`

and have docstrings to explain whats going on better